sphinx/domains/__init__.py

```diff
@@ -150,6 +150,8 @@ class Domain(object):
     indices = []            # type: List[Type[Index]]
     #: role name -> a warning message if reference is missing
     dangling_warnings = {}  # type: Dict[unicode, unicode]
+    #: node_class -> (enum_node_type, title_getter)
+    enumerable_nodes = {}   # type: Dict[nodes.Node, Tuple[unicode, Callable]]
```

Comment:
I'd like to discuss who should manage enumerable nodes. In this PR, I moved it to Domain class.
But I'm not sure it is responsibility of Domain or not, and I can't say this is best.
If you have any idea, please let me know.

Reply:
I have no objection to this implementation. It is reasonable to handle the enumeration of nodes belonging to a math domain within that domain.

One concern is, implementations that use this mechanism may not appear because the enumeration node does not come up with the existing "figure", "table", "code" and "math".